File: scripts/solar/solar-common/tool/STimer.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 6,["8"] = 6,["11"] = 9,["12"] = 11,["13"] = 11,["14"] = 11,["16"] = 14,["17"] = 14,["18"] = 14,["19"] = 15,["20"] = 16,["21"] = 17,["22"] = 18,["23"] = 19,["24"] = 20,["25"] = 21,["26"] = 22,["27"] = 23,["28"] = 24,["29"] = 26,["30"] = 44,["31"] = 45,["32"] = 46,["34"] = 48,["35"] = 44,["36"] = 52,["37"] = 53,["40"] = 56,["43"] = 59,["44"] = 61,["45"] = 63,["46"] = 64,["47"] = 65,["48"] = 66,["49"] = 67,["50"] = 68,["52"] = 71,["54"] = 74,["55"] = 74,["56"] = 75,["57"] = 74,["58"] = 74,["59"] = 74,["61"] = 79,["62"] = 52,["63"] = 109,["64"] = 110,["65"] = 109,["66"] = 113,["67"] = 114,["68"] = 115,["69"] = 113,["70"] = 121,["71"] = 122,["72"] = 123,["73"] = 121,["74"] = 129,["75"] = 130,["76"] = 131,["77"] = 132,["78"] = 133,["79"] = 129,["80"] = 139,["81"] = 140,["82"] = 141,["83"] = 142,["84"] = 143,["85"] = 144,["86"] = 145,["87"] = 139,["88"] = 154,["89"] = 155,["90"] = 156,["91"] = 157,["93"] = 159,["94"] = 160,["95"] = 161,["97"] = 164,["98"] = 166,["99"] = 168,["100"] = 154,["101"] = 178,["102"] = 179,["103"] = 180,["104"] = 181,["105"] = 182,["106"] = 183,["109"] = 186,["110"] = 178,["111"] = 189,["112"] = 190,["113"] = 191,["114"] = 192,["115"] = 193,["116"] = 194,["118"] = 196,["119"] = 189,["120"] = 199,["121"] = 200,["122"] = 201,["123"] = 202,["124"] = 203,["126"] = 206,["128"] = 199,["133"] = 34,["141"] = 41,["150"] = 86,["152"] = 94,["153"] = 95,["154"] = 96,["155"] = 97,["156"] = 98,["159"] = 102,["160"] = 103,["165"] = 176});
local ____exports = {}
local ____ErrorMsgHelper = require("solar.solar-common.common.ErrorMsgHelper")
local ErrorMsgHelper = ____ErrorMsgHelper.default
---
-- @noSelf *
local index = 0
____exports.default = __TS__Class()
local STimer = ____exports.default
STimer.name = "STimer"
function STimer.prototype.____constructor(self)
    local ____index_0 = index
    index = ____index_0 + 1
    self.uuid = ____index_0
    self._elapsed = 0
    self.interval = 0
    self._callCount = 0
    self.loopCount = 1
    self.isPause = false
    self._timeout = 0
    self._isDestroy = false
    self.periodic = false
    self.lastUpdateTime = 0
    self.handlerFunc = nil
end
function STimer.prototype.setLoopCount(self, loopCount)
    if loopCount == 0 then
        log.errorWithTraceBack("不能设置循环次数为0!")
    end
    self.loopCount = loopCount
end
function STimer.prototype.update(self)
    if self._isDestroy then
        return
    end
    if self.isPause then
        return
    end
    local tpf = _g_time - self.lastUpdateTime
    self.interval = self.interval + tpf
    if self.interval >= self._timeout then
        self._callCount = self._callCount + 1
        self.interval = 0
        self._elapsed = self._elapsed + self._timeout
        if not self.periodic and self._callCount >= self.loopCount then
            self:destroy()
        else
            ____exports.default:_sl_addSTimer2TimeCallBack(self._timeout, self)
        end
        xpcall(
            function()
                self.handlerFunc(self)
            end,
            ErrorMsgHelper.error_handle
        )
    end
    self.lastUpdateTime = _g_time
end
function STimer.prototype.isDestroy(self)
    return self._isDestroy
end
function STimer.prototype.destroy(self)
    self._isDestroy = true
    return self
end
function STimer.prototype.pause(self)
    self.isPause = true
    return self
end
function STimer.prototype.resume(self)
    self.isPause = false
    self:reset()
    ____exports.default:_sl_addSTimer2TimeCallBack(self._timeout, self)
    return self
end
function STimer.prototype.reset(self)
    self._elapsed = 0
    self.interval = 0
    self._callCount = 0
    self.isPause = false
    self._isDestroy = false
    self.lastUpdateTime = 0
end
function STimer.prototype.start(self, timeoutS, handlerFunc, periodic)
    self.handlerFunc = handlerFunc
    if periodic ~= nil then
        self.periodic = periodic
    end
    if timeoutS == 0 then
        handlerFunc(self)
        return self
    end
    self._timeout = math.floor(timeoutS * 100 + 0.5) * 10
    ____exports.default:_sl_addSTimer2TimeCallBack(self._timeout, self)
    return self
end
function STimer._sl_getSTimerCount(self)
    local count = 0
    for timeCallBackKey in pairs(____exports.default.timeCallBack) do
        local callBackArray = ____exports.default.timeCallBack[timeCallBackKey]
        if callBackArray then
            count = count + #callBackArray
        end
    end
    return count
end
function STimer._sl_addSTimer2TimeCallBack(self, timeout, sTimer)
    local callBackTime = _g_time + timeout
    local callBackArray = ____exports.default.timeCallBack[callBackTime]
    if callBackArray == nil then
        callBackArray = {}
        ____exports.default.timeCallBack[callBackTime] = callBackArray
    end
    callBackArray[#callBackArray + 1] = sTimer
end
function STimer._sl_OnTime()
    local callBackArray = ____exports.default.timeCallBack[_g_time]
    if callBackArray then
        for ____, callBackArrayElement in ipairs(callBackArray) do
            callBackArrayElement:update()
        end
        ____exports.default.timeCallBack[_g_time] = nil
    end
end
__TS__SetDescriptor(
    STimer.prototype,
    "elapsed",
    {get = function(self)
        return self._elapsed
    end},
    true
)
__TS__SetDescriptor(
    STimer.prototype,
    "remaining",
    {get = function(self)
        return self._timeout - self._elapsed
    end},
    true
)
__TS__SetDescriptor(
    STimer.prototype,
    "timeout",
    {
        get = function(self)
            return self._timeout
        end,
        set = function(self, timeMS)
            timeMS = math.floor(timeMS / 10 + 0.5) * 10
            if timeMS < 10 then
                log.errorWithTraceBack("计时器间隔不得低于0.01毫秒")
                self:destroy()
                return
            end
            self.lastUpdateTime = -timeMS
            self._timeout = timeMS
        end
    },
    true
)
STimer.timeCallBack = {}
return ____exports

File: src/solar/solar-common/tool/STimer.ts
/**
 * 太阳计时器
 */
/** @noSelfInFile **/

import ErrorMsgHelper from "@/ErrorMsgHelper";

/** @noSelf **/
let index = 0;

export default class STimer implements Destroyable {


    private readonly uuid: number = index++;
    private _elapsed: number = 0;
    private interval: number = 0;
    private _callCount: number = 0;
    private loopCount: number = 1;
    private isPause: boolean = false;
    private _timeout: number = 0;
    private _isDestroy: boolean = false;
    private periodic: boolean = false;
    private lastUpdateTime: number = 0;
    private handlerFunc: (this: void, timer: STimer) => void = null;

    constructor() {

    }

    /**
     * 计时器经过的时间
     */
    public get elapsed(): number {
        return this._elapsed;
    }

    /**
     * 计时器剩余时间
     */
    public get remaining(): number {
        return this._timeout - this._elapsed;
    }

    setLoopCount(loopCount: number) {
        if (loopCount == 0) {
            log.errorWithTraceBack("不能设置循环次数为0!")
        }
        this.loopCount = loopCount;
    }

    //毫秒单位
    update() {
        if (this._isDestroy) {
            return
        }
        if (this.isPause) {
            return
        }
        let tpf = _g_time - this.lastUpdateTime;
        //秒单位
        this.interval += tpf;
        //计算间隔逻辑回调
        if (this.interval >= this._timeout) {
            this._callCount++;
            this.interval = 0;
            this._elapsed += this._timeout;
            if (!this.periodic && this._callCount >= this.loopCount) {//不是循环的自动销毁
                this.destroy();
            } else {
                //加入下一次系统回调
                STimer._sl_addSTimer2TimeCallBack(this._timeout, this)
            }
            //即使报错也不要影响后续的逻辑 以尽可能提高游戏体验
            xpcall(() => {
                this.handlerFunc(this);
            }, ErrorMsgHelper.error_handle);

        }
        this.lastUpdateTime = _g_time;
    }

    /**
     * 计时器
     */
    public get timeout(): number {
        return this._timeout;
    }

    /**
     * 设置间隔 毫秒单位
     * (修改时 会在下次计时器到期时才能开始以 timeMS 的频率正常运行 即修改后的第一次运行可能会额外等0-1倍的之前的timeMS的延迟)
     * @param timeMS
     */
    set timeout(timeMS: number) {
        timeMS = Math.round(timeMS / 10) * 10;
        if (timeMS < 10) {
            log.errorWithTraceBack("计时器间隔不得低于0.01毫秒")
            this.destroy()
            return
        }
        //重新设置上次更新时间 以保证下次更新时能 到期时间满足条件
        this.lastUpdateTime = -timeMS;
        this._timeout = timeMS;


    }


    public isDestroy(): boolean {
        return this._isDestroy;
    }

    public destroy() {
        this._isDestroy = true;
        return this;
    }

    /**
     * @deprecated 暂不支持
     */
    public pause() {
        this.isPause = true;
        return this;
    }

    /**
     * @deprecated 暂不支持
     */
    public resume() {
        this.isPause = false;
        this.reset();
        STimer._sl_addSTimer2TimeCallBack(this._timeout, this);
        return this;
    }

    /**
     * 重设计时器 以便可以再次start
     */
    reset() {
        this._elapsed = 0;
        this.interval = 0;
        this._callCount = 0;
        this.isPause = false;
        this._isDestroy = false;
        this.lastUpdateTime = 0;
    }

    /**
     * 启动计时器
     * @param timeoutS  （单位：秒 精度为小数点后2位 0.01）
     * @param handlerFunc
     * @param periodic
     */
    public start(timeoutS: number, handlerFunc: (this: void, timer: STimer) => void, periodic?: boolean) {
        this.handlerFunc = handlerFunc;
        if (periodic != null) {
            this.periodic = periodic;
        }
        if (timeoutS == 0) {
            handlerFunc(this)
            return this;
        }
        // 0.024秒 设置为 20ms
        this._timeout = Math.round(timeoutS * 100) * 10;
        //
        STimer._sl_addSTimer2TimeCallBack(this._timeout, this);
        //
        return this;
    }


    /**
     * 全局静态
     * 注意这个数据是异步的
     */
    private static timeCallBack: { [time: number]: STimer[] } = {};

    static _sl_getSTimerCount(): number {
        let count = 0;
        for (let timeCallBackKey in STimer.timeCallBack) {
            let callBackArray: STimer[] = STimer.timeCallBack[timeCallBackKey];
            if (callBackArray) {
                count += callBackArray.length;
            }
        }
        return count;
    }

    private static _sl_addSTimer2TimeCallBack(timeout: number, sTimer: STimer) {
        let callBackTime = _g_time + timeout;
        let callBackArray: STimer[] = STimer.timeCallBack[callBackTime];
        if (callBackArray == null) {
            callBackArray = []
            STimer.timeCallBack[callBackTime] = callBackArray;
        }
        callBackArray.push(sTimer)
    }

    static _sl_OnTime(this: void) {
        let callBackArray = STimer.timeCallBack[_g_time];
        if (callBackArray) {
            for (let callBackArrayElement of callBackArray) {
                callBackArrayElement.update();
            }
            //删除旧的
            STimer.timeCallBack[_g_time] = null;
        }
    }

}
